File: C/src/CiftTaktik.c
/** 
* @file ./src/CiftTaktik.cpp
* @description CiftTaktik yapısının kaynak dosyası
* @course 1. öğretim B grubu
* @assignment 2. Ödev
* @date 18.05.2023
* @author Sude Çakmak sude.cakmak1@ogr.sakarya.edu.tr
*/

#include "CiftTaktik.h"

CiftTaktik CiftTakOlustur(int miktar){  //kurucu fonksiyon
	CiftTaktik this;
	this = (CiftTaktik)malloc(sizeof(struct CIFTTAKTIK));
	this->super = new_Taktik(miktar);   //kalıtım alınan yapı
	this->super->savas = &Savas;
	this->super->kalitimyoket = &CiftTakYoket; 
	this->yoket = &CiftTakYoket;
	return this;
}
int Savas(const Taktik this, int aktif_populasyon){  // 1 ile 1000 arası değer döndür
	int deger = rand() + (this->baslangic_populasyon+aktif_populasyon)*2;
	deger = abs(deger);
	deger = deger%1000;
		if (deger <= 1) return 2;  //0 ya da 1 ihtimaline karşı
		else return deger;
}
void CiftTakYoket(const CiftTaktik this){  //yıkıcı fonksiyon
	if(this == NULL) return;
	this->super->yoket(this->super);
	free(this);
}

File: Java/src/odev/CiftTaktik.java
/**
*
* @author Sude Çakmak / sude.cakmak1@ogr.sakarya.edu.tr
* @since 30.05.2023
* <p>
* Taktik için kullanılan sınıflardan biri (1-1000 arası rastgele değer döndürür)
* </p>
*/
package odev;

import java.util.Random;

public class CiftTaktik extends Taktik{    //Taktik soyut sınıfından kalıtım alıyor
	public CiftTaktik(int miktar) {
		super(miktar);   //soyut sınıfın yapıcısını çağırıyor
	}
	
	@Override  //soyut sınıftaki metodun gerçeklemesi
	public int Savas() {      // Koloniler savaşırken rastgele taktik değeri döndürerek savaşı kazandırır/kaybettirir
		Random r = new Random();       //dönecek olan değere rastgelelik katmak için
		int rastgele = this.baslangic_populasyon*2;   //bu metoda özgü bir şey olsun diye
		int rand = r.nextInt(2000);
		rastgele += rand;
		rastgele = (rastgele % 1000) ;  //0'dan 1000'e kadar bir değer (0 dahil 1000 dahil değil)
		if(1 < rastgele && rastgele < 1000) return rastgele;
		else return 4;  //0 ya da 1 gelme ihtimaline karşılık (1-1000 arasında sayı istendiği için 1 ve 1000 dahil değil)
	}
}
